linalg/multivector.hpp

```diff
@@ -0,0 +1,145 @@
+
+   /// Reorder Vector \p v from its current ordering \p in_ord to \p out_ord
+   static void Reorder(Vector &v, int vdim, Type in_ord, Type out_ord);
+};
+
+template <> inline int
+Ordering::Map<Ordering::byNODES>(int ndofs, int vdim, int dof, int vd)
+{
+   MFEM_ASSERT(dof < ndofs && -1-dof < ndofs && 0 <= vd && vd < vdim, "");
+   return (dof >= 0) ? dof+ndofs*vd : dof-ndofs*vd;
+}
+
+template <> inline int
+Ordering::Map<Ordering::byVDIM>(int ndofs, int vdim, int dof, int vd)
+{
+   MFEM_ASSERT(dof < ndofs && -1-dof < ndofs && 0 <= vd && vd < vdim, "");
+   return (dof >= 0) ? vd+vdim*dof : -1-(vd+vdim*(-1-dof));
+}
+
+
+/// MultiVector carries data for an arbitrary number of Vectors of a given size/vdim.
+class MultiVector : public Vector
+{
+protected:
+
+   /// Vector dimension.
+   int vdim;
+
+   /// Ordering of Vector data in MultiVector.
+   Ordering::Type ordering;
+
+   /// Re-allocate + copy memory. See Array::GrowSize.
+   void GrowSize(int min_num_vectors);
+
+public:
+
+   using Vector::operator=;
+   using Vector::operator();
+
+   MultiVector() : vdim(0), ordering(Ordering::byNODES) {};
+
+   /// Initialize an empty MultiVector of vdim \p vdim_ with ordering \p ordering_.
+   MultiVector(int vdim_, Ordering::Type ordering_);
+
+   /// Initialize a MultiVector with \p num_vectors vectors each of size \p vdim_ ordered \p ordering_.
+   MultiVector(int vdim_, Ordering::Type ordering_, int num_vectors);
+
+   /// Initialize a MultiVector of vdim \p vdim_ with ordering \p ordering_ , initialized with copy of data in \p vec .
+   MultiVector(int vdim_, Ordering::Type ordering_, const Vector &vec);
+
+   /// Get the Vector dimension of the MultiVector.
+   int GetVDim() const { return vdim; }
+
+   /// Get the ordering of data in the MultiVector.
+   Ordering::Type GetOrdering() const { return ordering; }
+
+   /// Get the number of Vectors in the MultiVector.
+   int GetNumVectors() const { return Size()/vdim; }
+
+   /// Get a copy of Vector \p i 's data.
+   void GetVectorValues(int i, Vector &nvals) const;
+
+   /** @brief For `GetOrdering` == Ordering::byVDIM, set \p nref to refer to Vector \p i 's data.
+    *
+    *  @warning This method only works when ordering is Ordering::byVDIM, where an individual Vector's data is stored contiguously in memory.
+    */
+   void GetVectorRef(int i, Vector &nref);
+
+   /// Set Vector \p i 's data to \p nvals .
+   void SetVectorValues(int i, const Vector &nvals);
+
+   /// Reference to Vector \p i component \p comp value.
+   real_t& operator()(int i, int comp);
+
+   /// Const reference to Vector \p i component \p comp value.
+   const real_t& operator()(int i, int comp) const;
+
+   /** @brief Remove Vector s at \p indices.
+    *
+    *  @details The MultiVector is resized appropriately.
+    */
+   void DeleteVectorsAt(const Array<int> &indices);
+
+   /// Set the vector dimension of the MultiVector.
+   void SetVDim(int vdim_) { vdim = vdim_; }
```

Comment:
I think we should resize and copy the data based on ordering if necessary?

Reply:
I'm not quite sure I'm following... Are you saying if for example vdim=3, but then `SetVDim(4)` is called, that we keep the previous data, but allocate/resize/shift so that there is now space for 4 components, default-initialized?